lib/puppet/type/cisco_bgp_af_aa.rb

```diff
@@ -0,0 +1,155 @@
+# limitations under the License.
+
+Puppet::Type.newtype(:cisco_bgp_af_aa) do
+  @doc = "Manages a cisco BGP Address family aggregate address
+
+    cisco_bgp_af_aa { '<title>':
+      ..attributes..
+    }
+
+    `<title>` is the title of the bgp_af_aa resource
+
+Example:
+
+    cisco_bgp_af_aa{ '55.77 default ipv4 multicast 1.1.1.1/32':
+      ensure  => 'present',
+      as_set   => 'true',
+      summary_only   => 'false',
+      advertise_map   => 'adm',
+      attribute_map   => 'atm',
+      suppress_map   => 'sum',
+    }
+  "
+
+  ###################
+  # Resource Naming #
+  ###################
+
+  # Parse out the title to fill in the attributes in these
+  # patterns. These attributes can be overwritten later.
+  def self.title_patterns
+    identity = ->(x) { x }
+    [
+      [
+        /^(\d+|\d+\.\d+) (\S+) (\S+) (\S+) (\S+)$/,
+        [
+          [:asn, identity],
+          [:vrf, identity],
+          [:afi, identity],
+          [:safi, identity],
+          [:aa, identity],
+        ],
+      ]
+    ]
+  end
+
+  ##############
+  # Parameters #
+  ##############
+
+  ensurable
+
+  # Overwrites the name method which by default returns only
+  # self[:name].
+  def name
+    "#{self[:asn]} #{self[:vrf]} #{self[:afi]} #{self[:safi]} #{self[:aa]}"
+  end
+
+  # Only needed to satisfy name parameter.
+  newparam(:name) do
+  end
+
+  newparam(:asn, namevar: true) do
+    desc "BGP autonomous system number.  Valid values are String, Integer in
+          ASPLAIN or ASDOT notation"
+    validate do |value|
+      unless /^(\d+|\d+\.\d+)$/.match(value.to_s)
+        fail("BGP asn #{value} must be specified in ASPLAIN or ASDOT notation")
+      end
+    end
+
+    munge(&:to_s)
+  end
+
+  newparam(:vrf, namevar: true) do
+    desc 'BGP vrf name. Valid values are string. ' \
+      "The name 'default' is a valid VRF."
+
+    defaultto('default')
+    newvalues(/^\S+$/)
+  end
+
+  newparam(:afi, namevar: true) do
+    desc 'BGP Address-family AFI (ipv4|ipv6). Valid values are string.'
+    newvalues(:ipv4, :ipv6)
+  end
+
+  newparam(:safi, namevar: true) do
+    desc 'BGP Address-family SAFI (unicast|multicast). Valid values are string.'
+    newvalues(:unicast, :multicast)
+  end
+
+  newparam(:aa, namevar: true) do
+    desc 'BGP aggregate address prefix. Valid values are String.'
+    munge(&:to_s)
+  end
+
+  ##############
+  # Attributes #
+  ##############
+
+  newproperty(:as_set) do
+    desc "Generates AS-SET information. Valid values are true, false or 'default'"
+
+    newvalues(:true, :false, :default)
+  end # property as_set
+
+  newproperty(:summary_only) do
+    desc "Stops advertising more specifics.  Valid values are true, false or 'default'"
+
+    newvalues(:true, :false, :default)
+  end # property summary_only
+
+  newproperty(:advertise_map) do
+    desc "Route map to select attribute information from specific routes. Valid values are String or 'default'"
+
+    munge do |value|
```

Comment:
Use the more concise syntax for these:

https://github.com/cisco/cisco-network-puppet-module/blob/2f1d60e358b7009bfedcf6175357603c9e115831/lib/puppet/type/cisco_bfd_global.rb#L92